sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -274,12 +275,30 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
   }
 
   object CartesianProduct extends Strategy {
+    def getSmallSide(left: LogicalPlan, right: LogicalPlan): BuildSide = {
+      if (right.statistics.sizeInBytes < left.statistics.sizeInBytes) {
+        joins.BuildRight
+      } else {
+        joins.BuildLeft
+      }
+    }
+
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
+      // If plan can broadcast we use BroadcastNestedLoopJoin, as we know for inner join with true
+      // condition is same as Cartesian.
+      case logical.Join(CanBroadcast(left), right, joinType, condition) =>
+        execution.joins.BroadcastNestedLoopJoin(
+          planLater(left), planLater(right), joins.BuildLeft, joinType, condition) :: Nil
+      case logical.Join(left, CanBroadcast(right), joinType, condition) =>
+        execution.joins.BroadcastNestedLoopJoin(
+          planLater(left), planLater(right), joins.BuildRight, joinType, condition) :: Nil
       case logical.Join(left, right, _, None) =>
-        execution.joins.CartesianProduct(planLater(left), planLater(right)) :: Nil
+        execution.joins.CartesianProduct(planLater(left), planLater(right),
+          getSmallSide(left, right)) :: Nil
       case logical.Join(left, right, Inner, Some(condition)) =>
         execution.Filter(condition,
-          execution.joins.CartesianProduct(planLater(left), planLater(right))) :: Nil
+          execution.joins.CartesianProduct(planLater(left), planLater(right),
+            getSmallSide(left, right))) :: Nil
```

Comment:
instead of passing a `BuildSide` to `CartesianProduct`, why not just change the parameters order according to the data size? like
```
if (left < right) {
  CartesianProduct(left, right)
} else {
  CartesianProduct(right, left)
}
```